crates/matrix-sdk-base/src/store/traits.rs

```diff
@@ -1163,6 +1239,42 @@ pub struct QueuedEvent {
     pub is_wedged: bool,
 }
 
+/// The specific user intent that characterizes a [`DependentQueuedEvent`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum DependentQueuedEventKind {
+    /// The event should be edited.
+    Edit {
+        /// The new event for the content.
+        new_content: SerializableEventContent,
+    },
+
+    /// The event should be redacted/aborted/removed.
+    Redact,
+}
+
+/// An event to be sent, depending on a [`QueuedEvent`] to be sent first.
+///
+/// Depending on whether the event has been sent or not, this will either update
+/// the local echo in the storage, or send an event equivalent to the user
+/// intent to the homeserver.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct DependentQueuedEvent {
+    /// Unique identifier for this dependent queued event.
+    ///
+    /// Useful for deletion.
+    pub id: usize,
```

Comment:
Hmm, there's a misunderstanding here, so that'll need more comments :)

- N dependent queued events may relate to a single queued event (e.g. a message I've sent can have an edit, and a reaction (later))
- the transaction id here is that of the related queued event, not one uniquely identifying the dependent queued event
- the `id` is what uniquely identifies the dependent queued event

I thought of adding a transaction id for the dependent queued event itself (that was the point behind the mysterious comment I made above regarding maybe needing another migration), but it turns out that only dependent events that are reactions need this, and the transaction id may be folded into the `DependentQueuedEventKind::Reaction` field I'm adding later. 

1. Should we use a transparent newtype wrapper for the id here?
2. I'll make sure the comments expand on the difference between transaction id and id here.
3. Should we also have a unique transaction id used for sending the dependent queued event itself? (in addition to the one making the connection with the original queued event)

Reply:
1. Sounds like a good idea.
2. Nice, thanks.
3. Well you need a unique transaction ID, the question is rather if you should save it as part of the dependent queued event struct? If so, it sounds like a good idea to me but I won't contest if you disagree.

Reply:
It might be better to include the transaction id there, at least to avoid redundant sending to the server, I'll add that later 😞